spec/fixtures/balancer_utils.lua

```diff
@@ -0,0 +1,533 @@
+    local path = "/upstreams/" .. upstream_id .."/health?balancer_health=1"
+    local status, body = api_send("GET", path, nil, forced_port)
+    if status == 200 then
+      return body
+    end
+  end
+
+  post_target_address_health = function(upstream_id, target_id, address, mode, forced_port)
+    local path = "/upstreams/" .. upstream_id .. "/targets/" .. target_id .. "/" .. address .. "/" .. mode
+    return api_send("POST", path, {}, forced_port)
+  end
+
+  get_router_version = function(forced_port)
+    local path = "/cache/router:version"
+    local status, body = api_send("GET", path, nil, forced_port)
+    if status == 200 then
+      return body.message
+    end
+  end
+
+  do
+    local os_name
+    do
+      local pd = io.popen("uname -s")
+      os_name = pd:read("*l")
+      pd:close()
+    end
+    local function port_in_use(port)
+      if os_name ~= "Linux" then
+        return false
+      end
+      return os.execute("netstat -n | grep -q -w " .. port)
+    end
+
+    local port = FIRST_PORT
+    gen_port = function()
+      repeat
+        port = port + 1
+      until not port_in_use(port)
+      return port
+    end
+  end
+
+  do
+    local host_num = 0
+    gen_multi_host = function()
+      host_num = host_num + 1
+      return "multiple-hosts-" .. tostring(host_num) .. ".test"
+    end
+  end
+
+  add_target = function(bp, upstream_id, host, port, data)
+    port = port or gen_port()
+    local req = utils.deep_copy(data) or {}
+    req.target = req.target or utils.format_host(host, port)
+    req.weight = req.weight or 10
+    req.upstream = { id = upstream_id }
+    bp.targets:insert(req)
+    return port
+  end
+
+  add_api = function(bp, upstream_name, opts)
+    opts = opts or {}
+    local route_id = utils.uuid()
+    local service_id = utils.uuid()
+    local route_host = gen_sym("host")
+    local sproto = opts.service_protocol or opts.route_protocol or "http"
+    local rproto = opts.route_protocol or "http"
+
+    bp.services:insert({
+      id = service_id,
+      url = sproto .. "://" .. upstream_name .. ":" .. (rproto == "tcp" and 9100 or 80),
+      read_timeout = opts.read_timeout,
+      write_timeout = opts.write_timeout,
+      connect_timeout = opts.connect_timeout,
+      retries = opts.retries,
+      protocol = sproto,
+    })
+    bp.routes:insert({
+      id = route_id,
+      service = { id = service_id },
+      protocols = { rproto },
+      hosts = rproto ~= "tcp" and { route_host } or nil,
+      destinations = (rproto == "tcp") and {{ port = 9100 }} or nil,
+    })
+    return route_host, service_id, route_id
+  end
+
+  patch_api = function(bp, service_id, new_upstream, read_timeout)
+    bp.services:update(service_id, {
+      url = new_upstream,
+      read_timeout = read_timeout,
+    })
+  end
+end
+
+
+local poll_wait_health
+local poll_wait_address_health
+do
+  local function poll_wait(upstream_id, host, port, admin_port, fn)
+    local hard_timeout = ngx.now() + 70
```

Comment:
Can this be rewritten using [`wait_until`](https://github.com/Kong/kong/blob/next/spec/helpers.lua#L315) instead? The code below seems to do exactly what `wait_until` does.